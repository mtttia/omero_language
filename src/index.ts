import { Generator } from "./compiler/Generator";
import { ICompileOptions } from "./compiler/Generator/ICompileOptions";
import { Lexer } from "./compiler/Lexer"
import { Parser } from "./compiler/Parser";

export function compileToJs(code: string, options: ICompileOptions): string
{
    var l = new Lexer(code);
    let tokens = l.tokenize();

    var p = new Parser(tokens);
    let parses = p.parse();

    var g = new Generator(parses, options)
    let generated = g.generate();

    return generated
}