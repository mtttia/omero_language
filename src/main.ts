import { Generator } from "./compiler/Generator";
import { ICompileOptions } from "./compiler/Generator/ICompileOptions";
import { Lexer } from "./compiler/Lexer"
import { Parser } from "./compiler/Parser";

const code = `
My dear GOD Ares you are now Oracle prophecy 1

My dear GOD Zeus you are now (((30 allied with 40) without 10) times 10) schism 5
My dear GOD Apollo you are now Ares

Duel Zeus Kill Apollo
Paride give golden apple
Proposal Afrodite
    Omero: Here the on true code
Proposal Atena
    Omero: Here the on false code
Troia Destruction

Titanomachy
    Titanomachy Or Loose
        Duel Zeus Kill (Oracle prophecy 2 allied with 5) without Apollo
        Duel Loose Zeus Die both with Ares
    Gods won
    Duel Or Zeus Fight with 10
Gods won
Penelope Weave the web
    Omero: here go the code
Ulisse back to Itaca

goddess say Zeus Omero: print the number
oh goddess tell me Zeus Omero: convert to the correspondent char
`

const defaultOptions: ICompileOptions = {
    arguments: ["1"],
    useInputArgs: false
}

var l = new Lexer(code);
let tokens = l.tokenize();

var p = new Parser(tokens);
let parses = p.parse();

var g = new Generator(parses, defaultOptions)
let generated = g.generate();

console.log(generated);