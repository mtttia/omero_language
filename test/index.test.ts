import { compileToJs } from "../src"

test("Simple program compile to JS", () =>
{
    const compiled = compileToJs(`My dear GOD Ares you are now Oracle prophecy 1

My dear GOD Zeus you are now (((30 allied with 40) without 10) times 10) schism 5
My dear GOD Apollo you are now Ares

Duel Zeus Kill Apollo
Paride give golden apple
Proposal Afrodite
    Omero: Here the on true code
Proposal Atena
    Omero: Here the on false code
Troia Destruction

Titanomachy
    Titanomachy Or Loose
        Duel Zeus Kill (Oracle prophecy 2 allied with 5) without Apollo
        Duel Loose Zeus Die both with Ares
    Gods won
    Duel Or Zeus Fight with 10
Gods won
Penelope Weave the web
    Omero: here go the code
Ulisse back to Itaca

goddess say Zeus Omero: print the number
oh goddess tell me Zeus Omero: convert to the correspondent char`, {
        arguments: ["1", "2"],
        useInputArgs: false
    })

    expect(compiled).toMatchSnapshot();
})

test("Simple program compile to JS with \\r\\n", () =>
{
    const compiled = compileToJs("My dear GOD Ares you are now Oracle prophecy 1\r\n\r\nMy dear GOD Zeus you are now (((30 allied with 40) without 10) times 10) schism 5\r\nMy dear GOD Apollo you are now Ares\r\n\r\nDuel Zeus Kill Apollo\r\nParide give golden apple\r\nProposal Afrodite\r\n    Omero: Here the on true code\r\nProposal Atena\r\n    Omero: Here the on false code\r\nTroia Destruction\r\n\r\nTitanomachy\r\n    Titanomachy Or Loose\r\n        Duel Zeus Kill (Oracle prophecy 2 allied with 5) without Apollo\r\n        Duel Loose Zeus Die both with Ares\r\n    Gods won\r\n    Duel Or Zeus Fight with 10\r\nGods won\r\nPenelope Weave the web\r\n    Omero: here go the code\r\nUlisse back to Itaca\r\n\r\ngoddess say Zeus Omero: print the number\r\noh goddess tell me Zeus Omero: convert to the correspondent char", {
        arguments: ["1", "2"],
        useInputArgs: false
    })

    expect(compiled).toMatchSnapshot()
})