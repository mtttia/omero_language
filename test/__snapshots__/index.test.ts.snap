// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simple program compile to JS 1`] = `
"let Ade = null;let Afrodite = null;let Apollo = null;let Ares = null;let Artemide = null;let Atena = null;let Demetra = null;let Dioniso = null;let Efesto = null;let Era = null;let Ermes = null;let Poseidone = null;let Zeus = null;let args = [1, 2]
Ares = args[0]  
Zeus = (((30 + 40) - 10) * 10) / 5
Apollo = Ares  

if(  ((Zeus  ) > (Apollo  ))){}else{}

while( ! (  ((Zeus  ) > ((args[1] + 5) - Apollo)) && ! ((Zeus  ) == (Ares  ))) ||  ((Zeus  ) >= (10  ))){}
console.log(Zeus  )
console.log(String.fromCharCode(Zeus  ))"
`;

exports[`Simple program compile to JS with \\r\\n 1`] = `
"let Ade = null;let Afrodite = null;let Apollo = null;let Ares = null;let Artemide = null;let Atena = null;let Demetra = null;let Dioniso = null;let Efesto = null;let Era = null;let Ermes = null;let Poseidone = null;let Zeus = null;let args = [1, 2]
Ares = args[0]  
Zeus = (((30 + 40) - 10) * 10) / 5
Apollo = Ares  

if(  ((Zeus  ) > (Apollo  ))){}else{}

while( ! (  ((Zeus  ) > ((args[1] + 5) - Apollo)) && ! ((Zeus  ) == (Ares  ))) ||  ((Zeus  ) >= (10  ))){}
console.log(Zeus  )
console.log(String.fromCharCode(Zeus  ))"
`;
